### DISCUSSION TITLE
is in substate use clone get  owner  ship is  a good  choice

### DISCUSSION THREAD

**Discussion by joebnb** [tokio-rs/axum#2499] [General]
```
use axum::{Router, routing::get, extract::{State, FromRef}};

// the application state
#[derive(Clone)]
struct AppState {
    // that holds some api specific state
    api_state: ApiState,
}

// the api specific state
#[derive(Clone)]
struct ApiState {}

// support converting an `AppState` in an `ApiState`
impl FromRef<AppState> for ApiState {
    fn from_ref(app_state: &AppState) -> ApiState {
        app_state.api_state.clone()
    }
}

let state = AppState {
    api_state: ApiState {},
};

let app = Router::new()
    .route("/", get(handler))
    .route("/api/users", get(api_users))
    .with_state(state);

async fn api_users(
    // access the api specific state
    State(api_state): State<ApiState>,
) {
}

async fn handler(
    // we can still access to top level state
    State(state): State<AppState>,
) {
}
```

im new bee of rust and  axum,im  my view clone is expensive,is  there a good way to get substate owner ship

**Comment by davidpdrsn** [comment]
It depends on what data is in your state. It it has a database connection pool then it’s fine. If it holds 2gb of binary data then it’s probably not fine.

**Reply by joebnb** [reply]
i want add memory session store in here it's seems a little bit tough,would you please share some parttens or some mind of good design?

this is my session store and i want use it in handler and middleware
![image](https://github.com/tokio-rs/axum/assets/1646883/6f1c7083-3a34-43fe-8786-a4478bceecb7)

**Reply by davidpdrsn** [reply]
I'd recommend you join [the tokio discord](https://discord.gg/tokio) and ask in `#axum`. There are more people there who can help :)
